nexus/src/app/sagas/instance_create.rs

```diff
@@ -299,11 +306,245 @@ impl NexusSaga for SagaInstanceCreate {
+                        *entry.first_port,
+                        *entry.last_port,
+                        &nat_target,
+                    )
+                    .await
+            }
+            ipnetwork::IpNetwork::V6(network) => {
+                dpd_client
+                    .nat_set_ipv6(
+                        &network.ip(),
+                        *entry.first_port,
+                        *entry.last_port,
+                        &nat_target,
+                    )
+                    .await
+            }
+        }
+        .map_err(|e| {
+            ActionError::action_failed(Error::internal_error(&format!(
+                "failed to create nat entry via dpd: {e}"
+            )))
+        })?;
+
+        debug!(log, "creation of nat entry successful for: {entry:#?}");
+    }
+    Ok(())
+}
+
+async fn sic_remove_network_config(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    let osagactx = sagactx.user_data();
+    let dpd_client = &osagactx.nexus().dpd_client;
+    let datastore = &osagactx.datastore();
+    let log = sagactx.user_data().log();
+
+    let instance_name = sagactx.lookup::<db::model::Name>("instance_name")?;
+    let (.., db_instance) = LookupPath::new(&opctx, &datastore)
+        .project_id(params.project_id)
+        .instance_name(&instance_name)
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let instance_id = db_instance.id();
+
+    debug!(log, "fetching external ip addresses");
+
+    let external_ips = &datastore
+        .instance_lookup_external_ips(&opctx, instance_id)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let mut entries_to_remove: Vec<ExternalIp> = vec![];
+
+    for entry in external_ips {
+        debug!(
+            log,
+            "checking for existing nat mapping for address: {entry:#?}"
+        );
+
+        let existing_nat = match entry.ip {
+            ipnetwork::IpNetwork::V4(network) => {
+                dpd_client.nat_get_ipv4(&network.ip(), *entry.first_port).await
+            }
+            ipnetwork::IpNetwork::V6(network) => {
+                dpd_client.nat_get_ipv6(&network.ip(), *entry.first_port).await
+            }
+        };
+
+        match existing_nat {
+            Ok(value) => {
+                // nat entry exists, we need to remove it
+                debug!(log, "found nat entry: {value:#?}");
+                entries_to_remove.push(entry.clone());
+            }
+            Err(_) => {
+                // TODO: correctness
+                // This needs a bit more logic since "not found" is an error,
+                // but in this case it is actually ok. We want to propogate
+                // other errors.
+            }
+        }
+    }
+
+    for entry in entries_to_remove {
```

Comment:
Same thing here about idempotency. We need a different forward and undo action for _each_ external IP address.

Reply:
Do we actually need a separate saga node for _each_ ip, or do we need to make sure that we attempt to delete the nat entry for every existing ip _at least once_?